File: katas/6 kyu/find-the-missing-letter/solution.py
import codewars_test as test
# -------------------------------- MY SOLUTION ------------------------------- #
def find_missing_letter(chars):
  for i in range(1,len(chars)):
    if ord(chars[i]) > ord(chars[i-1])+1:
      return chr(ord(chars[i])-1)

def find_missing_letter(chars):
  return [chr(ord(chars[i])-1) for i in range(1,len(chars)) if ord(chars[i])-1 != ord(chars[i-1])][0]
# ------------------------------ CLEVER SOLUTION ----------------------------- #
def find_missing_letter(chars):
  n = 0
  while ord(chars[n]) == ord(chars[n+1]) - 1:
      n += 1
  return chr(1+ord(chars[n]))

def find_missing_letter(c):
  return next(chr(ord(c[i])+1) for i in range(len(c)-1) if ord(c[i])+1 != ord(c[i+1]))
# ----------------------------------- TEST ----------------------------------- #
test.assert_equals(find_missing_letter(['a','b','c','d','f']), 'e')
test.assert_equals(find_missing_letter(['O','Q','R','S']), 'P')

File: katas/6 kyu/find-the-missing-letter/solution.php
<!-- find-the-missing-letter -->
<?php
use PHPUnit\Framework\TestCase;
/* ------------------------------- MY SOLUTION ------------------------------ */
function find_missing_letter(array $array): string {
  $c = ord($array[0]);
  foreach ($array as $chr) {
    if ($c == ord($chr)) {
      $c++;
    }
    else return chr($c);
  }
}
/* ----------------------------- CLEVER SOLUTION ---------------------------- */
function find_missing_letter(array $array): string {
  $alphabet = range($array[0], $array[count($array) - 1]);
  return array_values(array_diff($alphabet, $array))[0];
}
/* ---------------------------------- TEST ---------------------------------- */
class FindTheMissingLetterTest extends TestCase {
  public function testDescriptionExamples() {
    $this->assertEquals("e", find_missing_letter(['a','b','c','d','f']));
    $this->assertEquals("P", find_missing_letter(["O", "Q", "R", "S"]));
  }
}